File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int splitArray(vector<int>& nums, int k) {
        int start =0,end=0,mid=0,count=0,ans=0,sum=0;
        for(int i = 0; i < nums.size(); i++){
            start = max(start , nums[i]);
            end += nums[i];
        }

        while(start<= end){
            mid = start + (end-start)/2;
            count=1;sum=0;
            for(int i =0 ; i < nums.size(); i++){
               sum += nums[i];
               if(sum > mid){
                count++;
                sum = nums[i];
               }
            }

            if(count<=k){
                ans = mid;
                end = mid-1;
            }
            else{
                start = mid+1;
            }
        }
        return ans;
    }
};

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {


    boolean check(int[] nums, int k,int mid){
        int last = nums[0],p =1;
       for(int i=1;i<nums.length;i++){
        if(nums[i] > mid) return false;
          if((last + nums[i]) > mid){
            p++;
            last = nums[i];
          }
          else{
            last += nums[i];
          }
       }
       if(p<=k) return true;
       return false;
    }
    public int splitArray(int[] nums, int k) {
         
         int st = Integer.MIN_VALUE,end = 0,ans = 0,mid = 0;
         for(int i=0;i<nums.length;i++){
             st = Math.max(st,nums[i]);
             end += nums[i];
         }

         while(st<=end){
             mid = st + (end-st)/2;
             if(check(nums,k,mid)){
                 ans = mid;
                 end = mid -1;
             }
             else{
                st = mid +1; 
             }
         }
         return ans;
    }
}